lib/CL/devices/pthread/pthread_scheduler.c

```diff
@@ -442,18 +372,99 @@ pocl_pthread_prepare_kernel (void *data, _cl_command_node *cmd)
   pthread_scheduler_push_kernel (run_cmd);
 }
 
-static void
-pocl_pthread_exec_command (_cl_command_node *cmd,
-                           struct pool_thread_data *td)
+static _cl_command_node *
+check_cmd_queue_for_device (thread_data *td)
+{
+  _cl_command_node *cmd;
+  DL_FOREACH (scheduler.work_queue, cmd)
+  {
+    cl_device_id subd = cmd->device;
+    if (shall_we_run_this (td, subd))
+      {
+        DL_DELETE (scheduler.work_queue, cmd)
+        return cmd;
+      }
+  }
+
+  return NULL;
+}
+
+static kernel_run_command *
+check_kernel_queue_for_device (thread_data *td)
+{
+  kernel_run_command *cmd;
+  DL_FOREACH (scheduler.kernel_queue, cmd)
+  {
+    cl_device_id subd = cmd->device;
+    if (shall_we_run_this (td, subd))
+      return cmd;
+  }
+
+  return NULL;
+}
+
+static int
+pthread_scheduler_get_work (thread_data *td)
```

Comment:
This function needs comments on why it does what it does in a way that it avoids "troubles with subdevices". This won't be remembered after 6 months without documentation.

Reply:
OK